mlir/include/mlir/Dialect/MIOpen/gridwise_convolution_implicit_gemm_util.h

```diff
@@ -68,3 +69,140 @@ static void populateSeqVal(const ArrayAttr &seqAttr,
+  llvm::SmallVector<int64_t, 0> strideVal;
+  populateSeqVal(strideAttr, strideVal);
+
+  auto dilationAttr = op.getAttrOfType<ArrayAttr>("dilations");
+  llvm::SmallVector<int64_t, 0> dilationVal;
+  populateSeqVal(dilationAttr, dilationVal);
+
+  auto paddingAttr = op.getAttrOfType<ArrayAttr>("padding");
+  llvm::SmallVector<int64_t, 0> paddingVal;
+  populateSeqVal(paddingAttr, paddingVal);
+
+  return {opType, dimIndexVal, strideVal, dilationVal, paddingVal};
+}
+
+struct InitParamsNonXDL : InitParams {
+  InitParamsNonXDL(int64_t mPerBlock, int64_t nPerBlock, int64_t kPerBlock,
+                   int64_t mPerThread, int64_t nPerThread, int64_t bSize)
+      : InitParams{mPerBlock, nPerBlock, kPerBlock}, gemmMPerThread(mPerThread),
+        gemmNPerThread(nPerThread), blockSize(bSize) {}
+  int64_t gemmMPerThread;
+  int64_t gemmNPerThread;
+  int64_t blockSize;
+};
+
+class PopulateParams : public PopulateParamsBase {
+private:
+  llvm::SmallVector<InitParamsNonXDL, 4> initParameters = {
+      // M/block N/block K/block M/thread N/thread blockSize
+      {128, 128, 8, 4, 4, 256},
+      {128, 64, 8, 4, 4, 128},
+      {64, 128, 4, 4, 4, 128},
+      {32, 32, 4, 2, 2, 64},
+  };
+
+  LogicalResult
+  calculateGemmABlockCopyPerformanceParameters(InitParamsNonXDL *param,
+                                               ConvolutionContext &ctx,
+                                               DerivedParams &derived) {
+    return calculateInputDerivedParams(param, param->blockSize, ctx, true,
+                                       derived);
+  }
+
+  LogicalResult
+  calculateGemmBBlockCopyPerformanceParameters(InitParamsNonXDL *param,
+                                               ConvolutionContext &ctx,
+                                               DerivedParams &derived) {
+
+    return calculateInputDerivedParams(param, param->blockSize, ctx, false,
+                                       derived);
+  }
+
+  int64_t calculateGemmCDestDataPerWrite(ConvolutionContext &ctx) {
+    int64_t outputVecLen = 0;
+    if ((ctx.opType == miopen::ConvOpType::Conv2DOpType) &&
+        (ctx.dimIndexVal["ko"].first == 3)) {
+      // gemmM vectorizable. However, there is no parameters for vectorizing
+      // gemmM dimension for matrix C. Do nothing here.
+    } else if ((ctx.opType == miopen::ConvOpType::Conv2DBwdDataOpType) &&
+               (ctx.dimIndexVal["ci"].first == 3)) {
+      // gemmM vectorizable. However, there is no parameters for vectorizing
+      // gemmM dimension for matrix C. Do nothing here.
+    } else {
+      PopulateParams::obtainGemmCVecLen(ctx.opType, ctx.dimIndexVal,
+                                        outputVecLen);
+    }
+
+    if ((outputVecLen > 0) && (outputVecLen % 4 == 0)) {
+      return 4;
+    } else if ((outputVecLen > 0) && (outputVecLen % 2 == 0)) {
+      return 2;
+    }
+
+    return 1;
+  }
+
+public:
+  void paramsFromCtx(ConvolutionContext &ctx, InitParamsNonXDL &validParams,
```

Comment:
returns LogicalResult?

Reply:
I can but it will ended up always returning a `success()` result because this function *should* success. If it does not, it will hit at a few lines below at `llvm_unreachable` and bail out.

Reply:
Will do.